### PR TITLE
Provider interface / class structure for re-use

### LINKED ISSUE #27: Extract common and specific parts for provider
While working on this PR(https://github.com/councilofelders/opensignals/pull/16), @jrdi suggested to first do some refactoring on the provider to separate the generic and specific parts for a provider. 

The simplest structure I can think of is a base class provider which takes the current provider code, wraps it in a class but leaves one method unimplemented(namely the one method which does a call to the data provider get data for one ticker for a given period).

### PR DESCRIPTION
**Pull request by jaapterwoerds**
I separated the provider code in the part which is generic and a part which handles the interaction with Yahoo Finance API itself. I split up some large methods to satisfy the pylinter but didn't make any functional changes.

Other provider API might warrant some further refactoring:
- Use a bigger universe for training 
- Fetching data in batches
Closes #27

### CODE REVIEW

**Comment by jaapterwoerds** [comment]
@jrdi Your review would be appreciated.

**Comment by hellno** [comment]
What's the status on this? :)

**Comment by jaapterwoerds** [comment]
@jrdi @uuazed I hope you can find some time to get this merged so we can work on integrating some alternative price data provider in here 😸

File: src/opensignals/__main__.py [hunk 1]
```diff
@@ -24,8 +24,7 @@
 from docopt import docopt
 
 from opensignals import __version__
-from opensignals.data import yahoo
-from opensignals.features import RSI
+from opensignals.data.yahoo import Yahoo
```
Comment (uuazed) [inline src/opensignals/__main__.py:27]: I prefer importing only modules and packages, not individual classes. That makes it more obvious where something is coming from in the code. Also compare google*s python style guide with examples and why they prefer organizing imports that way: https://google.github.io/styleguide/pyguide.html#22-imports

File: src/opensignals/data/provider.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    @staticmethod
+    def get_tickers():
+        ticker_map = pd.read_csv(SIGNALS_TICKER_MAP)
+        ticker_map = ticker_map.dropna(subset=['yahoo'])
+        logger.info(f'Number of eligible tickers: {ticker_map.shape[0]}')
+
+        if ticker_map['yahoo'].duplicated().any():
+            num = ticker_map["yahoo"].duplicated().values().sum()
+            raise Exception(
+                f'Found duplicated {num}'
```
Comment (uuazed) [inline src/opensignals/data/provider.py:31]: I am sure that fits in one line

File: src/opensignals/data/provider.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+                ' yahoo tickers'
+            )
+
+        if ticker_map['bloomberg_ticker'].duplicated().any():
+            num = ticker_map["bloomberg_ticker"].duplicated().values().sum()
+            raise Exception(
+                f'Found duplicated {num}'
```
Comment (uuazed) [inline src/opensignals/data/provider.py:38]: should also fit in one line

File: src/opensignals/data/provider.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+        return ticker_map
+
+    @staticmethod
+    def get_ticker_data(db_dir):
+        ticker_data = pd.DataFrame({
+            'bloomberg_ticker': pd.Series([], dtype='str'),
+            'date': pd.Series([], dtype='datetime64[ns]')
+        })
+        if len(list(db_dir.rglob('*.parquet'))) > 0:
```
Comment (uuazed) [inline src/opensignals/data/provider.py:50]: should we prefix the parquet files somehow? To ensure these are opensignals files.

File: src/opensignals/data/provider.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+        num = ticker_data.bloomberg_ticker.unique().shape[0]
+        logger.info(f'Retrieving data for {num} tickers from the database')
+
+        return ticker_data
+
+    @staticmethod
+    def get_ticker_missing(ticker_data,
+                           ticker_map,
+                           last_friday=datetime.today() - relativedelta(weekday=FR(-1))):
```
Comment (uuazed) [inline src/opensignals/data/provider.py:61]: The statement defining the default value for `last_friday` is executed upon definition time, not at call time. That could eventually cause some weird behavior. The saver way is defaulting to None then add some `if last_friday is None: last_friday = ... `

File: src/opensignals/data/provider.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            ['bloomberg_ticker', 'yahoo', 'date_max']
+        ]
+
+        tickers_outdated['start'] = (
+                tickers_outdated['date_max'] + pd.DateOffset(1)
+        ).dt.strftime('%Y-%m-%d')
+        tickers_outdated.drop(columns=['date_max'], inplace=True)
+
+        return pd.concat(
+            [ticker_not_found, tickers_outdated]
```
Comment (uuazed) [inline src/opensignals/data/provider.py:92]: this should fit in one line

File: src/opensignals/data/provider.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+        )
+
+    @staticmethod
+    def get_live_data(ticker_data,  last_friday):
```
Comment (uuazed) [inline src/opensignals/data/provider.py:96]: double space. I wonder why the linter doesn't complain

Comment (uuazed) [inline src/opensignals/data/provider.py:96]: Please add type hints to all methods.

File: src/opensignals/data/provider.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        thursday_data = thursday_data[
+            ~thursday_data.bloomberg_ticker.isin(live_data.bloomberg_ticker.values)
+        ].copy()
+
+        live_data = pd.concat([live_data, thursday_data])
+        live_data = live_data.set_index('date')
+        return live_data
+
+    @staticmethod
+    def get_train_test_data(ticker_data, feature_names, targets):
```
Comment (uuazed) [inline src/opensignals/data/provider.py:115]: type hints

File: src/opensignals/data/provider.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+            ticker_data, targets,
+            on=['date', 'bloomberg_ticker'],
+            how='left'
+        )
+
+        logger.info(f'Found {ml_data.target.isna().sum()}'
+                    'rows without target, filling with 0.5')
+        ml_data['target'] = ml_data['target'].fillna(0.5)
+
+        # convert date to datetime and index on it
```
Comment (uuazed) [inline src/opensignals/data/provider.py:127]: comment doesn't apply, no conversion happening here

File: src/opensignals/data/provider.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        ml_data = ml_data[ml_data.index.weekday == 4]
+        # drop eras with under 50 observations per era
+        ml_data = ml_data[ml_data.index.value_counts() > 50]
+
+        # train test split
+        train_data = ml_data[ml_data['data_type'] == 'train']
+        test_data = ml_data[ml_data['data_type'] == 'validation']
+        return train_data, test_data
+
+    def get_data(self,
```
Comment (uuazed) [inline src/opensignals/data/provider.py:142]: type hints please

File: src/opensignals/data/provider.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+                 db_dir,
+                 features_generators=None,
+                 last_friday=datetime.today() - relativedelta(weekday=FR(-1)),
```
Comment (uuazed) [inline src/opensignals/data/provider.py:145]: as before, safer to compute that within the method.

File: src/opensignals/data/provider.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+            ticker_universe['bloomberg_ticker'])]
+
+        targets = pd.read_csv(SIGNALS_TARGETS)
+        targets['date'] = pd.to_datetime(
+            targets['friday_date'],
+            format='%Y%m%d'
+        )
+        targets['target'] = targets[target]
+
+        feature_names = []
```
Comment (uuazed) [inline src/opensignals/data/provider.py:166]: with the features prefix we might no longer need to explicitly collect the feature names, but I guess that is outside the scope of this PR

File: src/opensignals/data/provider.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+        # generate live data
+        live_data = Provider.get_live_data(ticker_data, last_friday)
+
+        return train_data, test_data, live_data, feature_names
+
+    def download_tickers(self, tickers, start):
+        start_epoch = int(datetime.strptime(start, '%Y-%m-%d').timestamp())
+        end_epoch = int(datetime.combine(date.today(), time()).timestamp())
+
+        pbar = tqdm(
```
Comment (uuazed) [inline src/opensignals/data/provider.py:182]: one line

**Review by uuazed — COMMENTED** [review]
Awesome, good job! 
I left some comments and will review again once you have addressed those.

File: src/opensignals/__main__.py (hunk 1 above)
Comment (jaapterwoerds) [reply src/opensignals/__main__.py:27]: Thanks for the link, I didn't knew the Google style guide. 

I personally find it better to import only the things you need. Especially in dynamic languages which have limited support for dead code detection and refactoring I thinks it's important to see why and which part of a library you're using a specific part in your code.

In this case I don't think it is that important, I'm happy to conform to whatever style is desired.

File: src/opensignals/data/provider.py (hunk 2 above)
Comment (jaapterwoerds) [reply src/opensignals/data/provider.py:31]: I'm sure too 😄.

File: src/opensignals/data/provider.py (hunk 4 above)
Comment (jaapterwoerds) [reply src/opensignals/data/provider.py:50]: I don't think we should asume there is a reason to mix the raw data files for opensignals with some other data used. The way I use this library is that I hav a folder the raw price data downloaded by opensignals (e.g price/yahoo) and for additional data  I have a similar structure <data type>/<data source>.

File: src/opensignals/data/provider.py (hunk 5 above)
Comment (jaapterwoerds) [reply src/opensignals/data/provider.py:61]: That's a good one. I think I did a faithful refactor, but this indeed is one of Python's amazing gotchas.

File: src/opensignals/data/provider.py (hunk 12 above)
Comment (jaapterwoerds) [reply src/opensignals/data/provider.py:166]: Agreed

File: src/opensignals/data/provider.py (hunk 9 above)
Comment (jaapterwoerds) [reply src/opensignals/data/provider.py:127]: Let's remove it indeed

**Comment by jaapterwoerds** [comment]
@uuazed I processed the comments and did some manual test. Please let me know what you think.

**Review by uuazed — APPROVED** [review]
lgtm

**Comment by uuazed** [comment]
Merged to master!

Does the code in README.md still work with that change?

**Comment by jaapterwoerds** [comment]
@uuazed Thanks! I'm running the README as we speak, so far so good 😃

**Comment by jaapterwoerds** [comment]
> @uuazed Thanks! I'm running the README as we speak, so far so good 😃

The run completed succesful locally 🥳
